pkg/cliplugins/workload/plugin/workload.go

```diff
@@ -0,0 +1,165 @@
+/*
+Copyright 2022 The KCP Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package plugin
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"time"
+
+	jsonpatch "github.com/evanphx/json-patch"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/clientcmd"
+
+	kcpclientset "github.com/kcp-dev/kcp/pkg/client/clientset/versioned"
+)
+
+// Cordon the workload cluster and mark it as unschedulable
+func (c *Config) Cordon(ctx context.Context, workloadClusterName string) error {
+	config, err := clientcmd.NewDefaultClientConfig(*c.startingConfig, c.overrides).ClientConfig()
+	if err != nil {
+		return err
+	}
+
+	err = modifyCordon(ctx, config, workloadClusterName, true)
+	if err != nil {
+		return err
+	}
+
+	fmt.Println("cordoned")
+	return nil
+}
+
+// Uncordon the workload cluster and mark it as schedulable
+func (c *Config) Uncordon(ctx context.Context, workloadClusterName string) error {
+	config, err := clientcmd.NewDefaultClientConfig(*c.startingConfig, c.overrides).ClientConfig()
+	if err != nil {
+		return err
+	}
+
+	err = modifyCordon(ctx, config, workloadClusterName, false)
+	if err != nil {
+		return err
+	}
+
+	fmt.Println("uncordoned")
+	return nil
+}
+
+// change the workload cluster cordon value
+func modifyCordon(ctx context.Context, config *rest.Config, workloadClusterName string, cordon bool) error {
+	kcpClient, err := kcpclientset.NewForConfig(config)
+	if err != nil {
+		return fmt.Errorf("failed to create kcp client: %w", err)
+	}
+
+	workloadCluster, err := kcpClient.WorkloadV1alpha1().WorkloadClusters().Get(ctx,
+		workloadClusterName,
+		metav1.GetOptions{},
+	)
+	if err != nil {
+		return fmt.Errorf("failed to get WorkloadCluster %s: %w", workloadClusterName, err)
+	}
+
+	// Get a JSON copy of the existing data
+	oldData, err := json.Marshal(workloadCluster)
+	if err != nil {
+		return fmt.Errorf("failed to marshal old data for WorkloadCluster %s: %w", workloadClusterName, err)
+	}
+
+	// make the changes needed to the resource
+	workloadCluster.Spec.Unschedulable = cordon
+
+	// if uncordon, make sure evictAfter is not set
+	if !cordon {
+		workloadCluster.Spec.EvictAfter = nil
+	}
+
+	// Get a JSON copy of the changed data
+	newData, err := json.Marshal(workloadCluster)
```

Comment:
If I understand this correctly, this will encode both the UID and resourceVersion in the patch, which will cause this command to fail when either change while it's running. For UID, that seems like what we want. What do we expect for UX on the resourceVersion? Do we want this command to fail if it races with some other status update or something? If not, we should put this into a retry loop or remove the resourceVersion from the patch.

Reply:
[Here](https://github.com/kcp-dev/kcp/blob/69aed93556a6729fcc8c5e11eed458b11603e0f8/pkg/reconciler/tenancy/clusterworkspaceshard/clusterworkspaceshard_controller.go#L165-L171) is an example of building an explicit new object to diff against. You'd just want to do it with the Spec.

Reply:
@stevekuznetsov I have submitted a change based on your comments